### ISSUE TITLE
Bio-Formats update scripts

### ISSUE THREAD

**Issue by sbesson** [fiji/fiji#156]
Hi @ctrueden,

unsure whether an issue is the best place to discuss the technical bits, feel free to redirect me to the ImageJ forum. There has been a few issues reported with the ImageJ Bio-Formats update site lately (http://forum.imagej.net/t/bioformats-micromanager-clash/3393, http://forum.imagej.net/t/recent-issue-with-bioformats-multiposition-file-opening-from-a-macro-macro/3496).

The component rearchitecture work in Bio-Formats 5.3.0 effectively broke some of the assumptions under the `update-bf.sh` script namely:

- the fact all components are in the `ome` org: low-level components are now uploaded to Maven Central under the `org.openmicroscopy` group
- the fact all components have the same version

For updating the Java 8 update site, my solution has been to derive a simple POM (https://github.com/sbesson/bio-formats-fiji) in order to update the JARs rather than hardcoding and duplicating the dependency tree in the Shell script. Is this a valid approach of do you see a better solution? Is this something we could re-use for fixing the Bio-Formats update site?

**Comment by ctrueden** [comment]
@sbesson Just want to make sure you saw this... it would be good to discuss the best way to proceed here. Again, I am sorry for my long silence before.

**Comment by sbesson** [comment]
> @sbesson Many apologies for the extremely long delay in reply. Unfortunately, I have still not overcome my inbox backlog since last fall.

Hi Curtis. no worries, we definitely understand how quickly backlog can build up.

> I agree that it is unfortunate that the Bio-Formats update site is now causing problems. (And indeed problems relating to it have cropped up all over, such as here.) I am sorry I have not had the bandwidth to work on it or even respond until now. 

My initial hope had been to maintain the existing scripts to account for the 5.3.0 architecture changes and open a PR against this repository. Unfortunately, I found no straightforward way to do this and I resolved this by migrating to the POM-based approach mentioned above.

> I do think the update site provides a critical service to users, so I would greatly appreciate if the site is maintained. Would the Bio-Formats team be OK with taking over the maintenance? I simply have too many other things going on to deal with it in a timely manner.
>>    My solution has been to derive a simple POM (https://github.com/sbesson/bio-formats-fiji) in order to update the JARs rather than hardcoding and duplicating the dependency tree in the Shell script. Is this a valid approach of do you see a better solution? Is this something we could re-use for fixing the Bio-Formats update site?

As a general comment, maintaining the Bio-Formats JARs for the Fiji and Java-8 update sites since Bio-Formats 5.1.5 has definitely been a useful experience. We have not had the occasion to discuss the maintenance question internally so far. Let me raise the topic and get back to you.

In the meantime, one general question about the impact of the service. I remember part of the context related to the initial set up of the Bio-Formats update site back in 2013 was to balance the speed of the Bio-Formats lifecycle which was still coupled to OMERO. Since then, much progress has been made to fasten the delivery of Bio-Formats releases especially when critical or external bug fixes are included. As we have not received or read any public request, are there some usage metrics for this site as opposed to the regular update sites shipping the release versions of Bio-Formats?

> Sounds great. I hope the solution is as simple as creating a new Jenkins job to utilize that repository, and plugging in the upload credentials I sent to you on Gitter via PM.

Thanks for the credentials. From a technical perspective, a first investigation step would likely be to look into the automation of the update and deployment process as this operation is still done manually for official releases.

**Comment by ctrueden** [comment]
> As we have not received or read any public request, are there some usage metrics for this site as opposed to the regular update sites shipping the release versions of Bio-Formats?

You can check the awstats and raw logs of `sites.imagej.net` for the `Bio-Formats` directory. @joshmoore has access to this, and knows how. If you do this, I suggest looking at the stats from before the breakage, since it is likely that usage dropped way down after the site became non-functional.

> much progress has been made to fasten the delivery of Bio-Formats releases especially when critical or external bug fixes are included.

I definitely appreciate that. But releases have still been on the order of one per month. I personally believe that it is vital for users to have an easy way to test changes before they are released—ideally as soon as they merge to the main integration branch. If the OME team decides they cannot maintain the Bio-Formats update site, then I will continue/resume doing so. Just let me know either way.

**Comment by ctrueden** [comment]
For what it's worth, the Bio-Formats update site is the third-most-popular non-core update site, narrowly behind Tboudier (3D suite) and ImageScience.

**Comment by sbesson** [comment]
>For what it's worth, the Bio-Formats update site is the third-most-popular non-core update site, narrowly behind Tboudier (3D suite) and ImageScience.

Many thanks for the extra information. 

> I definitely appreciate that. But releases have still been on the order of one per month. I personally believe that it is vital for users to have an easy way to test changes before they are released—ideally as soon as they merge to the main integration branch. If the OME team decides they cannot maintain the Bio-Formats update site, then I will continue/resume doing so. Just let me know either way.

Discussed the topic with the rest of the team yesterday and we certainly recognize the value of this update site. We will investigate the automation of uploading the latest JARs to an update site and hopefully the maintenance of the Bio-Formats update site. This will likely in the background task as we have other priorities but I would think we'll have a better idea of where we standup by the upcoming [OME Users meeting](https://www.openmicroscopy.org/site/community/minutes/meetings/12th-annual-users-meeting-2017).
Does that sound okay? We can keep using this issue until the update site is resolved to communicate if we encounter difficulties.

**Comment by ctrueden** [comment]
> Does that sound okay? We can keep using this issue until the update site is resolved to communicate if we encounter difficulties.

Sure, just keep me posted. In the meantime, the Bio-Formats update site will unfortunately remain broken.

**Comment by sbesson** [comment]
@ctrueden heads up on the status of this. Following the instructions on the Wiki (thanks to @hadim for the contribution), we configured Travis CI to deploy the latest SNAPSHOT JARs to the Bio-Formats update site (see https://travis-ci.org/openmicroscopy/bio-formats-fiji/builds/227941004). Additionally, a [daily cron job](https://docs.travis-ci.com/user/cron-jobs/) has been set up which should allow SNAPSHOTs to be uploaded to the ImageJ update site within 24h.

Closing this issue as this automation should fully address it.
